src/slice.rs

```
@@ -0,0 +1,192 @@
+///     let transfer = DmaTransfer::start(sub_slice);
+///     while !transfer.is_done() {}
+///     transfer.free()
+/// }
+///
+/// /// This function is bad because we cannot unsplit the slice.
+/// fn dma_transfer_bad2(buffer: &'static mut [u8], length: usize) -> &'static mut [u8] {
+///     let (slice_a, slice_b) = buffer.split_at_mut(length);
+///     let transfer = DmaTransfer::start(slice_a);
+///     while !transfer.is_done() {}
+///     let slice_a = transfer.free();
+///     // can't unsplit!!!
+///     slice_a
+/// }
+///
+/// /// This function is good—we can get the whole slice back!
+/// fn dma_transfer(buffer: &'static mut [u8], length: usize) -> &'static mut [u8] {
+///     let buffer_slice = buffer.into_buffer_slice(..length);
+///     let transfer = DmaTransfer::start(buffer_slice);
+///     while !transfer.is_done() {}
+///     let buffer_slice = transfer.free();
+///     buffer_slice.inner()
+/// }
+///
+/// const SIZE: usize = 1024;
+///
+/// let buffer = {
+///     static mut BUFFER: [u8; 1024] = [0_u8; SIZE];
+///     unsafe { &mut BUFFER }
+/// };
+///
+/// assert_eq!(buffer.len(), SIZE);
+///
+/// let returned_buffer = dma_transfer(buffer, 123);
+///
+/// assert_eq!(returned_buffer.len(), SIZE);
+/// ```
+pub struct BufferSlice<T> {
+    inner: T,
+    range: Range<usize>,
+}
+
+impl<T> BufferSlice<T> {
+    /// Create a new [BufferSlice]
+    fn new(inner: T, range: impl BufferRange) -> Self {
+        // The range must be span a non-negative length for internal logic to work
+        Self {
+            inner,
+            range: range.into_range(),
+        }
+    }
+
+    /// Consume the [BufferSlice] and return the wrapped value
+    pub fn inner(self) -> T {
+        self.inner
+    }
+
+    /// Coerce the given range into [0..len)
+    fn coerced_range(&self, len: usize) -> Range<usize> {
+        let start = self.range.start.min(len);
+        let end = self.range.end.min(len);
+        Range { start, end }
+    }
```

Wouldn't an incorrect range be a reason to panic instead of silently attempting to correct the mistake.

---

I agree. I changed the API to behave like the `get` method of a `slice`.

---

I'm not entirely sure what the most common use case will be, but I thought getting an invalid range always the user made a programming error. In that case it doesn't really make sense to have all these `Option<>` wrappers. I would have expected it to be behave like `buf[start..end]` which will panic if it's out of bounds.

I'm actually not sure what would be the most practical solution here. Maybe someone else should comment on this.

---

I mean there is a non-panicking `buf.get(start..end)` method and a panicking `buf[start..end]`. Perhaps we could have both? I am trying to emulate regular slices as much as possible, and I feel you could argue `buf.get(start..end)` isn't needed because it is always user error anyway. However, it might be useful in code which should not panic, or to more easily handle cases of overflows.

---

@eldruin I looked further into this. We would need four extension functions 
```
into_write_buffer_slice
into_write_buffer_slice_checked
into_read_buffer_slice_checked
into_read_buffer_slice
```
or something like that. This seems quite verbose compared to just having two that can accomplish the same thing with than added `.unwrap()`.

I am not sure a slice-like API is as easy as I thought it would be.